src/main.cpp

```diff
@@ -2177,75 +2383,153 @@ bool DisconnectBlock(CBlock& block, CValidationState& state, CBlockIndex* pindex
 
     bool fClean = true;
 
-    CBlockUndo blockUndo;
-    CDiskBlockPos pos = pindex->GetUndoPos();
-    if (pos.IsNull())
-        return error("DisconnectBlock(): no undo data available");
-    if (!UndoReadFromDisk(blockUndo, pos, pindex->pprev->GetBlockHash()))
-        return error("DisconnectBlock(): failure reading undo data");
+    CBlockUndo blockUndo;
+    CDiskBlockPos pos = pindex->GetUndoPos();
+    if (pos.IsNull())
+        return error("DisconnectBlock(): no undo data available");
+    if (!UndoReadFromDisk(blockUndo, pos, pindex->pprev->GetBlockHash()))
+        return error("DisconnectBlock(): failure reading undo data");
+
+    LogPrint("sc", "%s():%d - ===============> CBlockUndo red from DB:\n%s\n",
+        __func__, __LINE__, blockUndo.ToString());
+    // no coinbase in blockundo
+#if 0
+    if (blockUndo.vtxundo.size() + 1 != block.vtx.size() + block.vcert.size() )
+#else
+    if (blockUndo.vtxundo.size() < (block.vtx.size() - 1 + block.vcert.size()))
+#endif
+        return error("DisconnectBlock(): block and undo data inconsistent");
+
+    LogPrint("sc", "%s():%d - restoring sc coins if any\n", __func__, __LINE__);
+    if (!view.RestoreImmatureBalances(pindex->nHeight, blockUndo) )
```

Comment:
switch the order between  RestoreImmatureBalances and RevertCeasingScs to reflect ConnectBlock order

Reply:
Done